### ISSUE TITLE
EarleyParser sometimes gets confused when given two consecutive nonterminals that can parse overlapping tokens

### ISSUE THREAD

**Issue by vrthra** [uds-se/fuzzingbook#73]
```python
from fuzzingbook.Parser import EarleyParser

def main(grammar, inp):
    p = EarleyParser(grammar, canonical=True, start_symbol='<start>', log=True)
    for t in p.parse(inp):
        print(t)


mgrammar = {
    "<start>": [
        [
            "<ztwo>",
            "<empty>"
        ]
    ],
    "<empty>": [ [] ],
    "<zone>":  [ [], ["s"] ],
    "<ztwo>": [ [ "<zone>", "<zone>" ] ]
}

main(mgrammar, 's')
```
Expected: At least two parse trees; Obtained: only one

```
None chart[0]
<zone>:= |(0,0)
<empty>:= |(0,0)
<start>:= <ztwo> <empty> |(0,0)
<ztwo>:= <zone> <zone> |(0,0) 

s chart[1]
<zone>:= s |(0,1)
<ztwo>:= <zone> <zone> |(0,1)
<zone>:= |(1,1)
<empty>:= |(1,1)
<start>:= <ztwo> <empty> |(0,1) 

('<start>', [('<ztwo>', [('<zone>', []), ('<zone>', [('s', [])])]), ('<empty>', [])])
```

Simpler

```
mgrammar = {
    "<start>": [
        [
            "<ztwo>",
            "x"
        ]
    ],
    "<ztwo>": [ [ "<zone>", "<zone>" ] ],
    "<zone>":  [ ['ss'], ["s"] ],
}

main(mgrammar, 'sssx')
```

Or this one

```
mgrammar = {
    "<start>": [
        [
            "<ztwo>",
            "<empty>"
        ]
    ],
    "<ztwo>": [ [ "<zone>", "<zone>" ] ],
    "<zone>":  [ ['ss'], ["s"] ],
    "<empty>":  [[]],
}

main(mgrammar, 'sss')
```

Result
```
None chart[0]
 

s chart[1]
<zone>:= s |(0,1) 

s chart[2]
<zone>:= s s |(0,2)
<zone>:= s |(1,2)
<ztwo>:= <zone> <zone> |(0,2)
<empty>:= |(2,2)
<start>:= <ztwo> <empty> |(0,2) 

s chart[3]
<zone>:= s s |(1,3)
<zone>:= s |(2,3)
<ztwo>:= <zone> <zone> |(0,3)
<empty>:= |(3,3)
<start>:= <ztwo> <empty> |(0,3) 

('<start>', [('<ztwo>', [('<zone>', [('s', [])]), ('<zone>', [('ss', [])])]), ('<empty>', [])])
```
Expected: `(zone: ss, zone: s, empty:)` and `(zone: s, zone: ss, empty:)`

The problem goes away if one removes the last `<empty>` from `<start>`.
```
mgrammar = {
    "<start>": [
        [
            "<ztwo>",
        ]
    ],
    "<ztwo>": [ [ "<zone>", "<zone>" ] ],
    "<zone>":  [ ['ss'], ["s"] ],
}

main(mgrammar, 'sss')
```
Result
```
None chart[0]
 

s chart[1]
<zone>:= s |(0,1) 

s chart[2]
<zone>:= s s |(0,2)
<zone>:= s |(1,2)
<ztwo>:= <zone> <zone> |(0,2)
<start>:= <ztwo> |(0,2) 

s chart[3]
<zone>:= s s |(1,3)
<zone>:= s |(2,3)
<ztwo>:= <zone> <zone> |(0,3)
<start>:= <ztwo> |(0,3) 

('<start>', [('<ztwo>', [('<zone>', [('s', [])]), ('<zone>', [('ss', [])])])])
('<start>', [('<ztwo>', [('<zone>', [('ss', [])]), ('<zone>', [('s', [])])])])
```
Hypothesis: I do not think that the issue is in the parser itself as it can be seen that the states at each character locations are equivalent. Hence, some confusion in rebuilding the parse forest after parsing.

**Comment by vrthra** [comment]
`extract_trees` we currently have
```
 for p in zip(*ptrees):                                 
      results.append((name, p))
```

What this does is:
```
>>> a = [[1,2], [3,4], [5,6]]
>>> list(zip(*a))
[(1, 3, 5), (2, 4, 6)]
```
What we need is:
```
[(1,3,5), (1,3,6), (1,4,5), (1,4,6), (2,3,5), (2,3,6), (2,4,5), (2,4,6)] 
```

**Comment by vrthra** [comment]
Fixed in commit a57d1ac940c6bec78ba70f2e1ed04fd87aeeb49e
